[VarExporter] a new component to serialize values to plain PHP code

nicolas-grekas opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

This PR proposes moving what is currently the `PhpMarshaller` class in the Cache component to a separate component.

This component would provide only one public static method:
`VarExporter::export($value, bool &$isStaticValue = null): string`.

This method returns `$value` serialized as plain PHP code. Running this code creates the same exact data structure that `$value` contained. This is exactly like `serialize()` and `unserialize()`, from which all semantics are preserved (`__sleep`, `__wakeup` and `Serializable`).

The reason to use this method *vs* `serialize()` or even igbinary is performance: thanks to OPcache, the resulting code is significantly faster and more memory efficient than using `unserialize()` or `igbinary_unserialize()`.

Unlike `var_export()`, this works on any serializable PHP value.

It also provides a few improvements over `var_export()`/`serialize()`:
- the output is PSR-2 compatible
- the output can be re-indented without messing up with any `\r` or `\n` in the data
- missing classes throw a `ReflectionException` instead of being unserialized to a `PHP_Incomplete_Class` object
- references involving `SplObjectStorage`, `ArrayObject` or `ArrayIterator` instances are preserved
- `Reflection*`, `IteratorIterator` and `RecursiveIteratorIterator` classes throw an exception when being serialized (their unserialized version is broken anyway, see  https://bugs.php.net/76737.)

nicolas-grekas committed:
[VarExporter] a new component to serialize values to plain PHP code

tolry commented:
since the naming is similar/identical it might be a good idea to distinguish this from the php core function http://php.net/manual/de/function.var-export.php ?

nicolas-grekas commented:
@tolry we have the `VarDumper` component that provides a better `var_dump()` function. I think that the `VarExporter` name conveys what it should: this is a better `var_export()`.

tolry commented:
@nicolas-grekas I guessed so, just thought it might be easier to grasp, if the explanation mentioned var_export explicitely. Ah, I see it's in there now - did you add it or did I miss it before? :grin:

nicolas-grekas commented:
@tolry  yes, I updated the description.

PR is ready IMHO.

Here is what I added to the description above:
- the output is PSR-2 compatible
- the output can be re-indented without messing up with any `\r` or `\n` in the data
- missing classes throw a `ReflectionException` instead of being unserialized to a `PHP_Incomplete_Class` object
- references involving `SplObjectStorage`, `ArrayObject` or `ArrayIterator` instances are preserved
- `Reflection*`, `IteratorIterator` and `RecursiveIteratorIterator` classes throw an exception when being serialized (their unserialized version is broken anyway, see  https://bugs.php.net/76737.)

Koc commented on src/Symfony/Component/VarExporter/Internal/Exporter.php:
```diff
@@ -9,17 +9,17 @@
 /**
  * @author Nicolas Grekas <p@tchwork.com>
  *
  * @internal
```
Does this class stay internal after moving it to separate component?

sroze commented on src/Symfony/Component/VarExporter/VarExporter.php:
```diff
@@ -9,40 +9,38 @@
+use Symfony\Component\VarExporter\Internal\Values;
 
 /**
  * @author Nicolas Grekas <p@tchwork.com>
```
We usually put that at the end of the comment.

nicolas-grekas replied:
moved down

sroze commented on src/Symfony/Component/VarExporter/composer.json:
```diff
@@ -0,0 +1,36 @@
+    "name": "symfony/var-exporter",
+    "type": "library",
+    "description": "A blend of var_export() + serialize() to turn any serializable data structure to plain PHP code",
+    "keywords": ["export", "serialize"],
```
I'd add `symfony` and `var-exporter` as well, they are the component 💃

nicolas-grekas replied:
this is already in the name of the component, no need to duplicate

sroze commented on src/Symfony/Component/VarExporter/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+4.2.0
+-----
+
+ * added the component
```
```patch
-  * added the component
+ * Created the component (extracted `PhpMarshaller` from the `Cache` component)
```

Could be valuable to give a hint of where it comes from.

nicolas-grekas replied:
PhpMarshaller has never been released, better not give it any existence except in internal history
I'd keep as is :)

sroze reviewed (approved):
Except these two comments, it looks good to me 👌 

Note that it's very likely it will require more maintenance than when it was internal.

nicolas-grekas commented:
> Note that it's very likely it will require more maintenance than when it was internal.

That's why I added more tests for more cases ;)

Comments addressed, thanks for the review.

fabpot commented on src/Symfony/Component/VarExporter/README.md:
```diff
@@ -0,0 +1,33 @@
+=====================
+
+The VarExporter component allows exporting any serializable PHP data structure to
+plain PHP code. While doing so, it preserves all the semantics associated to the
```
associated with

fabpot commented on src/Symfony/Component/VarExporter/README.md:
```diff
@@ -0,0 +1,33 @@
+
+The VarExporter component allows exporting any serializable PHP data structure to
+plain PHP code. While doing so, it preserves all the semantics associated to the
+serialization mechanism of PHP (`__wakeup`, `__sleep`, `Serializable` esp.)
```
"esp" can be removed IMHO

fabpot commented on src/Symfony/Component/VarExporter/README.md:
```diff
@@ -0,0 +1,33 @@
+Unlike `var_export()`, this works on any serializable PHP value.
+
+It also provides a few improvements over `var_export()`/`serialize()`:
+ * the output is PSR-2 compatible
```
missing empty line before the list, and missing `;` at the end of each line but the last.

fabpot commented on src/Symfony/Component/VarExporter/README.md:
```diff
@@ -0,0 +1,33 @@
+   instances are preserved
+ * `Reflection*`, `IteratorIterator` and `RecursiveIteratorIterator` classes
+   throw an exception when being serialized (their unserialized version is broken
+   anyway, see  https://bugs.php.net/76737.)
```
dot outside the parenthesis

fabpot replied:
extra space after "see"

fabpot commented on src/Symfony/Component/VarExporter/VarExporter.php:
```diff
@@ -9,40 +9,50 @@
- * PhpMarshaller allows serializing PHP data structures using var_export()
- * while preserving all the semantics associated to serialize().
+ * VarExporter allows serializing PHP data structures to plain PHP code (like var_export())
+ * while preserving all the semantics associated to serialize() (unlike var_export()).
```
associated with

fabpot commented on src/Symfony/Component/VarExporter/VarExporter.php:
```diff
@@ -9,40 +9,50 @@
+     *
+     * @return string The value exported as PHP code
+     *
+     * @throw \Exception When the provided value cannot be serialized
```
@throws

nicolas-grekas commented:
Comments addressed thanks.

fabpot commented:
Thank you @nicolas-grekas.
